Sources/NIO/BaseSocketChannel.swift

```diff
@@ -469,32 +475,83 @@ class BaseSocketChannel<SocketType: BaseSocketProtocol>: SelectableChannel, Chan
         return try self.socket.remoteAddress()
     }
 
+    private final func flushNowAndRegisterIfNecessary(mayHaveReentered: Bool) {
+        assert(self.isOpen)
+        assert(self.lifecycleManager.isActive)
+
+        switch self.flushNowState {
+        case .inFlushNowCallingOut:
+            assert(mayHaveReentered)
+            // We're calling out from `flushNow` after `writeSocket`'s write loop has finished. We therefore need to
+            // replay the `flushNow` later because we otherwise might not write a flushed write.
+            self.eventLoop.execute {
+                self.flushNowAndRegisterIfNecessary(mayHaveReentered: false)
```

Comment:
I think this call here is exposing us to all kinds of harm, so we need to step back and rethink the runtime model here.

There are two assertions at the top of this function call: `assert(self.isOpen)` and `assert(self.lifecycleManager.isActive)`. Neither of those assertions can be guaranteed to hold after an event loop tick has occurred: arbitrary deregistrations and closures can occur entirely synchronously in between this current write operation and the next event loop tick.

We need to come to a deterministic strategy for `flushNow` to actually flush _now_. Otherwise we need to back right out and push the `flush` call into the next event loop tick.

Reply:
@Lukasa thanks! That's right

Reply:
will quickly add a test just so we're sure to have this.

Reply:
@Lukasa we need a method like `flushNowAndRegisterIfNecessary` though (possibly with a better name). `flushNow` is confusing as it doesn't actually do what `flush` does (setting a flush marker) but it tries to send as much flushed data as possible.